src/libraries/Microsoft.Extensions.Options/gen/Emitter.cs

```diff
@@ -0,0 +1,365 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+
+namespace Microsoft.Extensions.Options.Generators
+{
+    /// <summary>
+    /// Emits option validation.
+    /// </summary>
+    internal sealed class Emitter : EmitterBase
+    {
+        private const string StaticValidationAttributeHolderClassName = "__Attributes";
+        private const string StaticValidatorHolderClassName = "__Validators";
+        private const string StaticFieldHolderClassesNamespace = "__OptionValidationStaticInstances";
+        private const string StaticValidationAttributeHolderClassFQN = $"global::{StaticFieldHolderClassesNamespace}.{StaticValidationAttributeHolderClassName}";
+        private const string StaticValidatorHolderClassFQN = $"global::{StaticFieldHolderClassesNamespace}.{StaticValidatorHolderClassName}";
+        private sealed record StaticFieldInfo(string FieldTypeFQN, int FieldOrder, string FieldName, IList<string> InstantiationLines);
+
+        public string Emit(
+            IEnumerable<ValidatorType> validatorTypes,
+            CancellationToken cancellationToken)
+        {
+            var staticValidationAttributesDict = new Dictionary<string, StaticFieldInfo>();
+            var staticValidatorsDict = new Dictionary<string, StaticFieldInfo>();
+
+            foreach (var vt in validatorTypes.OrderBy(static lt => lt.Namespace + "." + lt.Name))
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                GenValidatorType(vt, ref staticValidationAttributesDict, ref staticValidatorsDict);
+            }
+
+            GenStaticClassWithStaticReadonlyFields(staticValidationAttributesDict.Values, StaticFieldHolderClassesNamespace, StaticValidationAttributeHolderClassName);
+            GenStaticClassWithStaticReadonlyFields(staticValidatorsDict.Values, StaticFieldHolderClassesNamespace, StaticValidatorHolderClassName);
+
+            return Capture();
+        }
+
+        private void GenValidatorType(ValidatorType vt, ref Dictionary<string, StaticFieldInfo> staticValidationAttributesDict, ref Dictionary<string, StaticFieldInfo> staticValidatorsDict)
+        {
+            OutLn("#pragma warning disable CS0618 // Type or member is obsolete");
+
+            if (vt.Namespace.Length > 0)
```

Comment:
Since we're assuming a newer C# compiler, instead of outputting a scoping namespace here, you can use a non-scoping namespace. This you let you remove the code at the bottom of this function which closes the namespace block, and would reduce the indent of all the generated code.

Reply:
I don't think we can do that because we generate a code under multiple namespaces in the same source file.